[backend] Add search terms

valeriocos opened the pull request:
This PR proposes to extend the metadata attributes by adding a new field (`search_terms`) with the purpose of simplifying query operations and avoding inspecting the data attribute. 

The search terms are included in a dict with the following shape:
```
        {
            'term-1': 'value-1',
            'term-2': 'value-2',
            'term-3': 'value-3',
        }
```
The search terms are added to the metadata information of each item in `search_terms` attribute. If search terms are not defined, `search_terms` will be set to `None`.

This new feature is showcased with the Jira backend.

valeriocos commented:
An initial evaluation of the new feature has been done to measure the impact on the raw indexes in GrimoireLab. The evaluation consisted in running the raw collection on a Jira with >10000 issues using 
the old and search terms approach, and then measure the size of the two indexes. The table below (derived using `_cat/indices`)  shows that around 32% of disk storage would be saved with the search terms approach.

| index        | docs.count | size |
| ------------- |:-------------:| -----:| 
| jira_search | 10796 | 73.8mb |
| jira_old | 10796 | 109.1mb |

Furthermore, the additional benefits of the search terms are:
- it's not needed anymore to index all information coming from a data source (but just the one at metadata level). See below an example of new mapping:
```
{
     "dynamic":true,
     "properties": {
         "data": {
             "dynamic":false,
             "properties": {}
            }
      }
}
```
- the raw index mapping in GrimoireLab would be homogeneous across all data sources (see below), thus opening the doors to new ways of storing raw data.
- the codebase of ELK would be reduced (mappings and `_fix_item` method would not be needed anymore).
- errors such immense term exceptions, too many attribute to index, mapping attribute inconsistencies would disappear, thus making the platform more robust.

valeriocos commented:
To use the search_terms in the filter raw a minor change is needed in the code [here](https://github.com/chaoss/grimoirelab-elk/blob/master/grimoire_elk/elastic_items.py#L35) (`FILTER_DATA_ATTR = 'search_terms.'`). Beyond this, the configurations are pretty the same:

```
"jira": [
            "https://xxx.jira.com --filter-raw=search_terms.project_key:ACME"
        ]
```

```
"jira": [
            "https://xxx.jira.com --filter-raw=data.fields.project.key:ACME"
        ]
```

valeriocos commented:
The PR is ELK is available at: https://github.com/chaoss/grimoirelab-elk/pull/665

sduenas commented on perceval/backend.py:
```diff
@@ -74,14 +74,27 @@ class Backend:
     Classified data filtering and archiving are not compatible to prevent
     data leaks or security issues.
 
+    Each backend can also provides a set of search terms to simplify query
```
I would changed this with:

```suggestion
    Each backend might provide a set of search terms to simplify query
```

valeriocos replied:
I'm not sure about the use of `might` (which expresses permission) over `can` (which expresses ability/capability) in this sentence.
I'll fix the typo (`provides`)

sduenas commented on perceval/backend.py:
```diff
@@ -74,14 +74,27 @@ class Backend:
+            'term-3': 'value-3',
+        }
+
+    The search terms are added to the metadata information of each
```
```suggestion
    These terms are added to the metadata information of each
```

sduenas commented on perceval/backend.py:
```diff
@@ -74,14 +74,27 @@ class Backend:
+        }
+
+    The search terms are added to the metadata information of each
+    Perceval item in `search_terms` attributes. If `search_terms` is not set,
```
```suggestion
    item in `search_terms` attribute. If `search_terms` is not set,
```

sduenas commented on perceval/backend.py:
```diff
@@ -253,6 +267,10 @@ def metadata_updated_on(item):
         raise NotImplementedError
 
+    @staticmethod
+    def search_terms(item):
```
Wouldn't it be better to use the same behaviour we have for `classified_fields`?.

We have a common method in `Backend` class that removes those fields. They are defined in `CLASSIFIED_FIELDS` class variable. I think it will be better to have something similar, so we just need a piece of code to copy these fields and then each backend would define which fields are going to be included.

valeriocos replied:
I will give it a try

sduenas commented on perceval/backends/core/jira.py:
```diff
@@ -216,6 +216,22 @@ def metadata_category(item):
+        project_name = item['fields']['project']['name']
+
+        terms = {
+            'project_id': project_id,
```
Does make sense to include the `id` of the item too? The one returned by `metadata_id()`.

sduenas commented on tests/test_jira.py:
```diff
@@ -328,6 +328,86 @@ def request_callback(method, uri, headers):
+        issues = [issue for issue in jira.fetch()]
+
+        issue = issues[0]
+        self.assertEqual(issue['uuid'], '6a7ba2a01aee56603b9d8a5f6b40c843fc089b2f')
```
To align this section with the other ones you should also add this line:

```
self.assertEqual(issue['origin'], 'http://example.com')
```

sduenas commented on perceval/backend.py:
```diff
@@ -253,6 +267,10 @@ def metadata_updated_on(item):
         raise NotImplementedError
 
+    @staticmethod
+    def search_terms(item):
```
Would it be better to use `search_fields` instead of `search_terms`? I'm saying this because we have `classified_fields`, so maybe it makes more sense to align it with that concept too.

valeriocos commented:
I have prepared another PR addressing your comments: https://github.com/chaoss/grimoirelab-perceval/pull/560, so we can better evaluate the two solutions.

valeriocos commented:
Closing this PR since PR #560 seems better
